Add support for Shift-Delete, Ctrl-Insert and Shift-Insert

logiclrd opened the pull request:
This PR adds `SingleActivator` mappings for IBM CUA style clipboard accessors, with which Shift-Delete, Ctrl-Insert and Shift-Insert are equivalent to ^X ^C ^V. These mappings are natively supported on Windows and Linux (but not OS X).

~Not sure what to do about:~
- ~Documentation: Are ^X ^C ^V already documented?~
- ~Tests: Are there existing tests for ^X ^C ^V already? Is it possible to mock out the clipboard so that a test of this functionality doesn't hit the actual system clipboard?~
- ~OS X: Is it a problem that, with this change, Flutter will accept these additional shortcuts even though they aren't a part of the platform paradigm?~

UPDATE:
I'm not aware of existing documentation of clipboard shortcuts.

I have added tests, both of the existing ^X ^C ^V and of the IBM CUA style mappings added by this PR, and in the current iteration the changes do not add IBM CUA style mappings on OS X.

Fixes: #178483

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

logiclrd committed:
Updated default_text_editing_shortcuts.dart:
- Added SingleActivator mappings for Shift-Delete, Ctrl-Insert and Shift-Insert (equivalent to ^X ^C ^V).
- Added comment blocks explaining the provenance.

logiclrd commented on packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart:
```diff
@@ -291,12 +291,29 @@ class DefaultTextEditingShortcuts extends StatelessWidget {
+    const SingleActivator(LogicalKeyboardKey.insert, control: true): CopySelectionTextIntent.copy,
+    const SingleActivator(LogicalKeyboardKey.insert, shift: true): const PasteTextIntent(
+      SelectionChangedCause.keyboard,
+    ),
```
Huh, I confess I hadn't even noticed the Shift-Delete "delete character forward" shortcut.

There must be a platform where this is standard, but now that I know it's there, if I go to use it, it feels _extremely weird_. I have never used a system that behaved that way before. 😛

logiclrd replied:
I'm going to push a commit with some tests. They're untested, because I can't run the tests with the version of Dart I have installed and the download page for the Dart development snapshots seems to be broken right now. (Whatever browser I use, the console shows:

```
ClientException: Too few arguments passed. Expected 2 or more, got 3 instead., uri=https://storage.googleapis.com/storage/v1/b/dart-archive/o?delimiter=%2F&prefix=channels%2Fdev%2Frelease%2F&alt=json
main.clients.mjs:73     at /main.clients.wasm
    at /main.clients.wasm
    at main.clients.mjs:263:86
    at _223 (main.clients.mjs:264:27)
    at /main.clients.wasm
    at /main.clients.wasm
    at /main.clients.wasm
    at /main.clients.wasm
```

).

I will continue to try to run tests locally, but the C.I. build will also be able to give me some feedback in the meantime.

logiclrd replied:
I figured out how to use `flutter channel` 🙂

logiclrd committed:
Removed the prior mapping of Shift-Delete to "delete character forward" in default_text_editing_shortcuts.dart.

logiclrd committed:
Added tests of clipboard shortcuts to default_text_editing_shortcuts_test.dart.

logiclrd committed:
Reintroduced the SingleActivator binding for an unmodified Delete to delete the character forward in default_text_editing_shortcuts.dart.

logiclrd committed:
Rework of the default text editing shortcuts:
- Moved the clipboard shortcuts into their own dedicated map _clipboardShortcuts.
- Updated _androidShortcuts, _linuxShortcuts and _windowsShortcuts to include _clipboardShortcuts.
- Updated _webDisablingTextShortcuts to generate disabling shortcuts dynamically for everything in _clipboardShortcuts, removing the explicit ^X, ^C and ^V.

logiclrd committed:
Added methods .and() and .except() to the TargetPlatformVariant class in widget_tester.dart to allow fluent syntax for configuring a specific set of target variants.
Updated clipboard shortcuts tests in default_text_editing_shortcuts_test.dart to target desktop except macOS, and android.
Updated ActionSpyState in default_text_editing_shortcuts_test.dart to capture the intent for CopySelectionTextIntent and PasteTextIntent.

logiclrd committed:
Switched the order of .except and .and calls in default_text_editing_shortcuts_test.dart for clarity.

logiclrd committed:
Per the "Linux analyze" build, added a "const" to the construction of a SingleActivator line moved out of the pressShift loop.

logiclrd committed:
Ran `dart format` on default_text_editing_shortcuts.dart and default_test_editing_shortcuts_test.dart.

bleroux commented on packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart:
```diff
@@ -290,13 +289,32 @@ class DefaultTextEditingShortcuts extends StatelessWidget {
+    // Xerox/Apple: ^X ^C ^V
+    // -> Standard on Windows
+    // -> Standard on Linux
+    // -> Standard on OS X (with Command as modifier)
```
```suggestion
    // -> Standard on macOS (with Command as modifier)
```

bleroux reviewed (approved):
LGTM! Thanks for this great contribution :pray: 

Some minor nits and see my comment about the fluent API.

logiclrd committed:
Changed "OS X" to "Mac OS X" in default_text_editing_shortcuts.dart to match similar references elsewhere in the codebase.
Removed the fluent TargetPlatformVariant syntax and replaced its uses with explicit sets.

loic-sharma reviewed (approved):
Thanks for the contribution!

loic-sharma committed:
Merge branch 'master' into classic-clipboard-key-combos

logiclrd committed:
Omit obvious local variable types in default_text_editing_shortcuts_test.dart.

justinmc reviewed (approved):
LGTM 👍 . Thanks for catching this. I'm always worried there are a few obscure keyboard shortcuts that Flutter is missing.

logiclrd committed:
Merge branch 'master' into classic-clipboard-key-combos
